[WTF] Fix behavioral issues in searching in spans

calvaris opened the pull request:
#### 647785e34521e96ba37dab6bbd61a29ff2172666
<pre>
[WTF] Fix behavioral issues in searching in spans
<a href="https://bugs.webkit.org/show_bug.cgi?id=302925">https://bugs.webkit.org/show_bug.cgi?id=302925</a>

Reviewed by Darin Adler.

There were several cases where we were not handling properly certain cases when inspecting spans for data and some
codepaths that had different behaviors. Fixed them and improved test coverage. reverseFind lost the starting offset that
was not used anywhere and for coherence with the find counterpart.

Test: Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp
* Source/WTF/wtf/StdLibExtras.h:
(WTF::find):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::findIgnoringASCIICase):
(WTF::reverseFindInner):
(WTF::reverseFind):
* Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:
(TestWebKitAPI::TEST(WTF_StringCommon, Equal)):
(TestWebKitAPI::TEST(WTF_StringCommon, EqualIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, StartsWith)):
(TestWebKitAPI::TEST(WTF_StringCommon, EndsWith)):
(TestWebKitAPI::TEST(WTF_StringCommon, Find)):
(TestWebKitAPI::TEST(WTF_StringCommon, ReverseFind)):
(TestWebKitAPI::TEST(WTF_StringCommon, Contains)):
(TestWebKitAPI::TEST(WTF_StringCommon, StartsWithLettersIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, EndsWithLettersIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, FindIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, ContainsIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, CharactersAreAllASCII)):

Canonical link: <a href="https://commits.webkit.org/303965@main">https://commits.webkit.org/303965@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8834ecb9bd338084650567b919eabcff546ab80

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133932 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141512 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85993 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135802 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6974 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6307 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102453 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69756 "Too many flaky failures: compositing/layer-creation/multiple-keyframes-individual-transform-properties-animation-overlap.html, compositing/layer-creation/scale-rotation-individual-properties-animation-overlap.html, compositing/layer-creation/translate-scale-individual-properties-animation-overlap.html, css3/unicode-bidi-float-crash.html, fast/css/content-visibility-zoom.html, fast/forms/input-number-click.html, fast/html/crash-marquee-fullscreen.html, fast/line-grid/line-align-right-edges.html, fast/rendering/missing-selection-with-sibling-table.html, http/tests/security/offscreen-canvas-remote-read-remote-image.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136879 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4904 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120068 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83252 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4781 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2398 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126009 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113942 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38188 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144157 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132446 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6113 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38767 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110820 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6195 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5189 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111028 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4635 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116326 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59860 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6165 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34584 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165409 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6011 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69629 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43198 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6256 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6119 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -522,7 +522,17 @@ template<typename SearchCharacterType, typename MatchCharacterType>
+    if (matchCharacters.empty())
+        return 0;
+
+    auto sum = checkedSum<size_t>(matchCharacters.size() + startOffset);
```
Using the addition operator and _then_ passing the result to `checkedSum()` cannot check for overflow, we want to have both parameters added to the template function:

```suggestion
    auto sum = checkedSum<size_t>(matchCharacters.size(), startOffset);
```

aperezdc commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -832,6 +842,18 @@ inline bool contains(std::span<const CharacterType> characters, ASCIILiteral mat
+    if (matchCharacters.empty())
+        return 0;
+
+    auto sum = checkedSum<size_t>(matchCharacters.size() + start);
```
Same here:


```suggestion
    auto sum = checkedSum<size_t>(matchCharacters.size(), start);
```

aperezdc commented on Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:
```diff
@@ -196,7 +239,11 @@ TEST(WTF_StringCommon, FindIgnoringASCIICase)
+    EXPECT_EQ(WTF::findIgnoringASCIICase(std::span<const char8_t>(), std::span<const char8_t>()), 0UZ);
+    // EXPECT_EQ(WTF::findIgnoringASCIICase(std::span<const char8_t>(), unsafeSpanChar8("🍉WaterMelon🍉")), notFound);
+    // EXPECT_EQ(WTF::findIgnoringASCIICase(""_spanChar8, unsafeSpanChar8("🍉WaterMelon🍉")), notFound);
+    // EXPECT_EQ(WTF::findIgnoringASCIICase(unsafeSpanChar8("🍉"), unsafeSpanChar8("🍉WaterMelon🍉")), notFound);
```
Were these lines left commented accidentally? If they are supposed to not compile, please add a comment mentioning that.

aperezdc commented on Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:
```diff
@@ -205,6 +252,10 @@ TEST(WTF_StringCommon, ContainsIgnoringASCIICase)
+    EXPECT_TRUE(WTF::containsIgnoringASCIICase(std::span<const char8_t>(), std::span<const char8_t>()));
+    // EXPECT_FALSE(WTF::containsIgnoringASCIICase(std::span<const char8_t>(), unsafeSpanChar8("🍉WaterMelon🍉")));
+    // EXPECT_FALSE(WTF::containsIgnoringASCIICase(""_spanChar8, unsafeSpanChar8("🍉WaterMelon🍉")));
+    // EXPECT_FALSE(WTF::containsIgnoringASCIICase(unsafeSpanChar8("🍉"), unsafeSpanChar8("🍉WaterMelon🍉")));
```
Ditto.

aperezdc reviewed (commented):
The added logic LGTM; but please check the comments about `checkedSum()`.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -522,7 +522,17 @@ template<typename SearchCharacterType, typename MatchCharacterType>
 {
-    ASSERT(source.size() >= matchCharacters.size());
+    if (matchCharacters.empty())
+        return 0;
```
I don’t think this is the best value to return if matchCharacters is empty. I think we want to return `std::min(startOffset, source.size())` rather than `0`. Please add test cases for this, and if you think `0` is better, please explain why.
```suggestion
    if (matchCharacters.empty())
        return std::min(startOffset, source.size());
```

But I might suggest that if `startOffset` is `> source.size()` we should return `notFound`, even if we are looking for the empty string. Easiest way to do that is probably to remove the special check for empty string entirely (see my suggestion below). But another way would be to write something like this:
```suggestion
    if (matchCharacters.empty())
        return startOffset > source.size() ? notFound : startOffset;
```

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -522,7 +522,17 @@ template<typename SearchCharacterType, typename MatchCharacterType>
+    }
+
+    if (difference.value() < matchCharacters.size())
+        return notFound;
```
I don’t think we should do this checking by preflighting. It’s too easy to get it wrong when the check is different from the code below that relies on the check. Instead I think we should incorporate the check into the code below, and take advantage of the clamping that is already done by the `subspan` function rather than redoing it.

We could either use `checkedDifference` to compute `delta`, or just write:
```
if (startSearchedCharacters.size() < matchCharacters.size())
    return notFound;
```

But an even better way to write this that I think would be clearer is to eliminate `startSearchedCharacters` and `delta` entirely, and not add an explicit check of the length:
```
for (size_t offset = startOffset; offset < source.size(); ++offset) {
    if (equalIgnoringASCIICaseWithLength(source.subspan(offset), matchCharacters, matchCharacters.size())
        return offset;
}
return notFound;
```

I don’t see any downside to writing it this more straightforward way, and if we write enough test cases we can make sure this new version is correct. I think we could also consider eliminating the explicit check for empty `matchCharacters` and just write this:

```
for (size_t offset = startOffset; offset <= source.size(); ++offset) {
    if (equalIgnoringASCIICaseWithLength(source.subspan(offset), matchCharacters, matchCharacters.size())
        return offset;
}
return notFound;
```

I’m pretty sure this handles edge cases like startOffset > source.size() correctly without adding any preflights or checked arithmetic, and I would prefer this version.

calvaris replied:
I think I managed to change it from "pre-flight" to "in-flight". All tests pass.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -832,6 +842,12 @@ inline bool contains(std::span<const CharacterType> characters, ASCIILiteral mat
 ALWAYS_INLINE static size_t reverseFindInner(std::span<const SearchCharacterType> searchCharacters, std::span<const MatchCharacterType> matchCharacters, size_t start)
 {
+    if (matchCharacters.empty())
+        return searchCharacters.size();
```
Why do we need to add this? What test fails if we don’t add it?

calvaris replied:
None.

calvaris replied:
I removed that part.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -858,9 +874,9 @@ ALWAYS_INLINE static size_t reverseFindInner(std::span<const SearchCharacterType
 template<typename SearchCharacterType, typename MatchCharacterType>
     requires(TriviallyComparableCodeUnits<SearchCharacterType, MatchCharacterType>)
-ALWAYS_INLINE static size_t reverseFind(std::span<const SearchCharacterType> searchCharacters, std::span<const MatchCharacterType> matchCharacters, size_t start = std::numeric_limits<size_t>::max())
+ALWAYS_INLINE static size_t reverseFind(std::span<const SearchCharacterType> searchCharacters, std::span<const MatchCharacterType> matchCharacters)
```
What’s the rationale for removing the “start offset” feature from the `reverseFind` function?

calvaris replied:
Consistence. find does not have such parameter.

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1070,13 +1070,13 @@ template<typename T, std::size_t TExtent, typename U, std::size_t UExtent>
 size_t find(std::span<T, TExtent> haystack, std::span<U, UExtent> needle)
 {
-#if !HAVE(MEMMEM)
     if (needle.empty())
         return 0;
```
Why is this check needed? This is already what `memmem` does when the needle size is zero. Did some test indicate that was not the case on some platform?

calvaris replied:
```
../../../Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:183
Expected equality of these values:
  WTF::find(std::span<const char8_t>(), std::span<const char8_t>())
    Which is: 18446744073709551615
  0UZ
    Which is: 0
```

and line 184, 221 and 222 .

calvaris replied:
Some of the lines changed with the new version of the patch.

darinadler replied:
I understand why; it’s a limitation of `memmem`’s interface. When the haystack pointer is `nullptr`, then we can’t tell “not found” apart from “found at start of string” for the result of `memem`, and so searching for a zero length string incorrectly results in “not found”.

darinadler commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1070,13 +1070,13 @@ template<typename T, std::size_t TExtent, typename U, std::size_t UExtent>
     if (needle.empty())
         return 0;
 
     if (haystack.size() < needle.size())
         return notFound;
```
Why is this check needed? This is already what `memem` does when the needle size is greater than the haystack size. Did some test indicate that was not the case on some platform?

calvaris replied:
It looks like I can remove this part and leave only the check above.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -494,16 +494,9 @@ template<typename SearchCharacterType, typename MatchCharacterType>
-    for (size_t i = 0; i <= delta; ++i) {
-        if (equalIgnoringASCIICaseWithLength(startSearchedCharacters.subspan(i), matchCharacters, matchCharacters.size()))
-            return startOffset + i;
+    for (size_t offset = startOffset; offset <= source.size() && (source.size() - offset) >= matchCharacters.size(); ++offset) {
```
I personally like it better without the parentheses:
```suggestion
   for (size_t offset = startOffset; offset <= source.size() && source.size() - offset >= matchCharacters.size(); ++offset) {
```

calvaris committed:
[WTF] Fix behavioral issues in searching in spans
https://bugs.webkit.org/show_bug.cgi?id=302925

Reviewed by Darin Adler.

There were several cases where we were not handling properly certain cases when inspecting spans for data and some
codepaths that had different behaviors. Fixed them and improved test coverage. reverseFind lost the starting offset that
was not used anywhere and for coherence with the find counterpart.

Test: Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp
* Source/WTF/wtf/StdLibExtras.h:
(WTF::find):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::findIgnoringASCIICase):
(WTF::reverseFindInner):
(WTF::reverseFind):
* Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:
(TestWebKitAPI::TEST(WTF_StringCommon, Equal)):
(TestWebKitAPI::TEST(WTF_StringCommon, EqualIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, StartsWith)):
(TestWebKitAPI::TEST(WTF_StringCommon, EndsWith)):
(TestWebKitAPI::TEST(WTF_StringCommon, Find)):
(TestWebKitAPI::TEST(WTF_StringCommon, ReverseFind)):
(TestWebKitAPI::TEST(WTF_StringCommon, Contains)):
(TestWebKitAPI::TEST(WTF_StringCommon, StartsWithLettersIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, EndsWithLettersIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, FindIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, ContainsIgnoringASCIICase)):
(TestWebKitAPI::TEST(WTF_StringCommon, CharactersAreAllASCII)):

Canonical link: https://commits.webkit.org/303965@main
